torch/distributed/elastic/utils/gpu_health_check.py

```diff
@@ -0,0 +1,382 @@
+        """
+        return {0: "mlx5_0", 1: "mlx5_1", 2: "mlx5_3", 3: "mlx5_4"}
+
+    @with_pynvml_lock
+    def get_gpu_pci_mapping(self):
+        """
+        Retrieve GPU local rank to PCI Bus mapping using pynvml.
+        """
+        assert (
+            self.pynvml is not None
+        ), "pynvml is not initialized. Ensure check_pynvml_availability() is called first."
+
+        gpu_pci_map = {}
+        try:
+            self.pynvml.nvmlInit()
+            num_gpus = self.pynvml.nvmlDeviceGetCount()
+
+            for i in range(num_gpus):
+                handle = self.pynvml.nvmlDeviceGetHandleByIndex(i)
+                pci_info = self.pynvml.nvmlDeviceGetPciInfo(handle)
+                bus_id = pci_info.busId
+                if isinstance(bus_id, bytes):
+                    bus_id = bus_id.decode('utf-8')
+                bus_id = bus_id.lower()
+                # Extract the last 12 characters (standard PCI format)
+                gpu_pci_map[i] = bus_id[-12:]
+
+        except self.pynvml.NVMLError as e:
+            logger.error(f"NVML Error: {e}\n{traceback.format_exc()}")
+
+        finally:
+            try:
+                self.pynvml.nvmlShutdown()
+            except self.pynvml.NVMLError as e:
+                logger.error(f"Failed to shut down NVML: {e}")
+
+        return gpu_pci_map
+
+
+class GPUHealthCheck(PynvmlMixin):
+    def __init__(
+        self,
+        device_index: Optional[int] = None,
+        interval: int = 60,
+        on_failure: Optional[Callable] = None,
+    ):
+        """
+        Initializes the GPUHealthCheck class.
+
+        Args:
+            device_index (Optional[int]): GPU device index to check. If None, checks all GPUs.
+            interval (int): Interval in seconds between asynchronous health checks.
+            on_failure (Optional[Callable]): Callback function to handle health check failures.
+        """
+        super().__init__()
+        self.device_index = device_index
+        self.interval = interval
+        self.on_failure = on_failure
+        self.pynvml_available = self.check_pynvml_availability()
+        self.enabled = self._check_driver_version()
+
+    @with_pynvml_lock
+    def _check_driver_version(self) -> bool:
+        """
+        Checks if the GPU driver version supports health checks (version r570 or newer).
+
+        Returns:
+            bool: True if the driver supports health checks, False otherwise.
+        """
+        GPU_RECOVERY_API_MIN_DRIVER_VERSION = 570
+
+        if not self.pynvml_available:
+            logger.warning("GPU Health checks are disabled because pynvml is not available.")
+            return False
+
+        try:
+            self.pynvml.nvmlInit()
+            driver_version = self.pynvml.nvmlSystemGetDriverVersion()
+            if isinstance(driver_version, bytes):
+                driver_version = driver_version.decode('utf-8')
+            self.pynvml.nvmlShutdown()
```

Comment:
Can we have some better way of managing this rather than manually putting in nvmlInit/shutdown calls? Maybe a context manager?